src/main/java/com/cloudant/search3/Search.java

```diff
@@ -292,11 +294,13 @@ public void close() {
     private <R> R execute(
             final Index index,
             final LuceneFunction<SearchHandler, R> f) throws IOException, ParseException, ExecutionException {
+        final SearchCacheKey key = new SearchCacheKey(index);
         try {
-            final SearchHandler handler = handlers.get(new SearchCacheKey(index));
+            final SearchHandler handler = handlers.get(key);
             return f.apply(handler);
         } catch (final IOException | AlreadyClosedException | ExecutionException e) {
-            failedHandler(index, e);
+            LOGGER.warn("Invalidating handler for index {} for reason {}.", toSubspace(index), e.getMessage());
+            handlers.invalidate(key);
```

Comment:
I see you were calling `handlers.get(new SearchCacheKey(index))` but then calling
`handlers.invalidate(index)`

Reply:
right. 'invalidate()' took an Object rather than a SearchCacheKey, so type safety didn't catch it.